File: scripts/tcp_client.py
#!/usr/bin/env python3
"""TCP client for testing Scarlet TCP server

Usage: python3 tcp_client.py <host> <port>
"""

import sys
import socket

def main():
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <host> <port>")
        print(f"Example: {sys.argv[0]} 10.0.2.15 8080")
        return 1

    host = sys.argv[1]
    port = int(sys.argv[2])

    print(f"[TCP Client] Connecting to {host}:{port}...")

    try:
        # Create TCP socket
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)

        # Connect to server
        sock.connect((host, port))
        print(f"[TCP Client] Connected to {host}:{port}")

        # Send a message
        message = "Hello from Python client!"
        sock.sendall(message.encode('utf-8'))
        print(f"[TCP Client] Sent: {message}")

        # Receive response
        data = sock.recv(1024)
        print(f"[TCP Client] Received: {data.decode('utf-8')}")

        # Close connection
        sock.close()
        print(f"[TCP Client] Connection closed")

        return 0

    except socket.error as e:
        print(f"[TCP Client] Connection failed: {e}")
        return 1

if __name__ == '__main__':
    sys.exit(main())

File: user/std-bin/src/tcp_client.rs
use std::env;
use std::io::{Read, Write};
use std::net::{Ipv4Addr, SocketAddrV4, TcpStream};
use std::process::ExitCode;

fn main() -> ExitCode {
    println!("[tcp-client] Rust std version");

    let args = env::args().collect::<Vec<_>>();
    if args.len() != 3 {
        println!("Usage: tcp_client <host> <port>");
        println!("Example: tcp_client 10.0.2.2 8080");
        return ExitCode::from(1);
    }

    let host = match args[1].parse::<Ipv4Addr>() {
        Ok(host) => host,
        Err(err) => {
            println!("[tcp-client] Invalid IPv4 address: {err}");
            return ExitCode::from(1);
        }
    };
    let port = match args[2].parse::<u16>() {
        Ok(port) if port != 0 => port,
        _ => {
            println!("[tcp-client] Invalid port number");
            return ExitCode::from(1);
        }
    };

    let addr = SocketAddrV4::new(host, port);
    println!("[tcp-client] Connecting to {addr}...");

    let mut stream = match TcpStream::connect(addr) {
        Ok(stream) => stream,
        Err(err) => {
            println!("[tcp-client] Connection failed: {err}");
            return ExitCode::from(1);
        }
    };

    println!("[tcp-client] Connected");

    let message = b"Hello from Scarlet TCP client!";
    match stream.write(message) {
        Ok(n) => println!("[tcp-client] Sent {n} bytes"),
        Err(err) => {
            println!("[tcp-client] Send failed: {err}");
            return ExitCode::from(1);
        }
    }

    let mut buf = [0; 1024];
    match stream.read(&mut buf) {
        Ok(0) => println!("[tcp-client] Server closed connection"),
        Ok(n) => println!("[tcp-client] Received {n} bytes"),
        Err(err) => {
            println!("[tcp-client] Receive failed: {err}");
            return ExitCode::from(1);
        }
    }

    println!("[tcp-client] Done");
    ExitCode::SUCCESS
}
